### PR TITLE
[IMP] stock_account: add lot valuation

### PR DESCRIPTION
**Pull request by Whenrow**
Ability to track valuation by lot or serial numbers. This is an optional setting on each product. The valuation layers will be linked to a stock.lot record. A stock move may generate multiple valuation layers if constituted of different lots among their stock move line.

Task : 3079384

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/stock_account/models/stock_valuation_layer.py [hunk 1]
```diff
@@ -192,3 +194,63 @@ def _consume_all(self, qty_valued, valued, qty_to_value):
             new_valuation = unit_cost * new_valued_qty
 
         return new_valued_qty, new_valuation
+
+    def _change_standart_price_accounting_entries(self, new_price):
+        # Handle account moves.
```
Comment (amoyaux) [inline addons/stock_account/models/stock_valuation_layer.py:199]: I would create correction layers instead of cheating

File: addons/stock_account/models/stock_lot.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+class StockLot(models.Model):
+    _inherit = 'stock.lot'
+
+    value_svl = fields.Float(compute='_compute_value_svl', compute_sudo=True)
+    quantity_svl = fields.Float(compute='_compute_value_svl', compute_sudo=True)
+    avg_cost = fields.Monetary(string="Average Cost", compute='_compute_value_svl', compute_sudo=True, currency_field='company_currency_id')
+    total_value = fields.Monetary(string="Total Value", compute='_compute_value_svl', compute_sudo=True, currency_field='company_currency_id')
+    company_currency_id = fields.Many2one('res.currency', 'Valuation Currency', compute='_compute_value_svl', compute_sudo=True)
+    stock_valuation_layer_ids = fields.One2many('stock.valuation.layer', 'lot_id')
+    standard_price = fields.Float(
```
Comment (amoyaux) [inline addons/stock_account/models/stock_lot.py:19]: should be a compute with inverse function no? In order to not store nor synchronize it

File: addons/stock_account/models/product.py [hunk 3]
```diff
@@ -14,6 +14,32 @@ class ProductTemplate(models.Model):
 
     cost_method = fields.Selection(related="categ_id.property_cost_method", readonly=True)
     valuation = fields.Selection(related="categ_id.property_valuation", readonly=True)
+    lot_valuated = fields.Boolean(
```
Comment (amoyaux) [inline addons/stock_account/models/product.py:17]: I understand that it could be enable only on a subset of products but I would still allow the option on the category

Comment (amoyaux) [reply addons/stock_account/models/product.py:17]: It's confusing to have a part of the configuration on different objects

File: addons/stock_account/models/stock_move_line.py [hunk 4]
```diff
@@ -48,7 +49,35 @@ def write(self, vals):
...
+            # remove quantity of old lot
+            for move_line in self:
+                if move_line.state != 'done':
+                    continue
+                move = move_line.move_id
+                rounding = move.product_id.uom_id.rounding
+                diff = move.product_uom._compute_quantity(move_line.quantity, move.product_id.uom_id, rounding_method='HALF-UP')
+                if float_is_zero(diff, precision_rounding=rounding):
+                    continue
+                self._create_correction_svl(move, -diff)
```
Comment (amoyaux) [inline addons/stock_account/models/stock_move_line.py:68]: Not possible to introduce it in the upper loop and introduce the concept of lot in _create_correction_svl

File: addons/stock_account/models/stock_move_line.py [hunk 5]
```diff
@@ -59,23 +88,30 @@ def unlink(self):
         analytic_move_to_recompute._account_analytic_entry_move()
         return res
 
+    def _action_done(self):
```
Comment (amoyaux) [inline addons/stock_account/models/stock_move_line.py:91]: So we don't handle svl without lot?

File: addons/stock_account/models/product.py [hunk 6]
```diff
@@ -72,6 +110,100 @@ def write(self, vals):
...
+        svl_vals_list = out_svl_vals_list
+        for variant in products:
+            rounding = variant.uom_id.rounding
+            if float_is_zero(variant.quantity_svl, precision_rounding=rounding):
+                continue
+            price_unit = variant.value_svl / variant.quantity_svl
+            for lot, qty in lot_by_product[variant].items():
+                qty_to_remove = min(qty, p_orig[variant.id])
+                p_orig[variant.id] -= qty_to_remove
+                svl_vals = variant._prepare_in_svl_vals(qty_to_remove, price_unit, lot_id=lot)
```
Comment (amoyaux) [inline addons/stock_account/models/product.py:148]: You should take the value on the lot if possible and fallback on product standard price

File: addons/stock_account/models/stock_lot.py (hunk 2 above)
Comment (amoyaux) [reply addons/stock_account/models/stock_lot.py:19]: Nope since we could use for inventory adjustement with import the lot and their value before doing the inventory

File: addons/stock_account/views/product_views.xml [hunk 7]
```diff
@@ -50,6 +50,17 @@
         </record>
 
         <!-- Stock Report View -->
+        <record model="ir.ui.view" id="view_template_property_form_stock_account">
+            <field name="name">view.template.property.form.stock.account</field>
+            <field name="model">product.template</field>
+            <field name="inherit_id" ref="stock.view_template_property_form"/>
+            <field name="arch" type="xml">
+                <xpath expr="//field[@name='product_tooltip']" position="after">
+                    <field name="lot_valuated" invisible="tracking == 'none'"/>
```
Comment (amoyaux) [inline addons/stock_account/views/product_views.xml:59]: should be in the accounting menu

File: addons/stock_account/models/product.py [hunk 8]
```diff
@@ -14,6 +14,32 @@ class ProductTemplate(models.Model):
 
     cost_method = fields.Selection(related="categ_id.property_cost_method", readonly=True)
     valuation = fields.Selection(related="categ_id.property_valuation", readonly=True)
+    lot_valuated = fields.Boolean(
+        "Valuation by Lot/Serial number",
+        help="If checked, the valuation will be specific by Lot/Serial number, the valuation method is expected to be average cost.",
```
Comment (amoyaux) [inline addons/stock_account/models/product.py:19]: Don't forget to update

File: addons/stock_account/models/product.py [hunk 9]
```diff
@@ -14,6 +14,32 @@ class ProductTemplate(models.Model):
...
+    )
+
+    @api.onchange('standard_price')
+    def _onchange_standard_price(self):
+        if self.lot_valuated:
```
Comment (amoyaux) [inline addons/stock_account/models/product.py:24]: Don't trigger the warning when there is no layer or empty stock

File: addons/stock_account/models/stock_lot.py [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+        "Cost", company_dependent=True,
+        digits='Product Price', groups="base.group_user",
+        help="""Value of the lot (automatically computed in AVCO).
+        Used to value the product when the purchase cost is not known (e.g. inventory adjustment).
+        Used to compute margins on sale orders."""
+    )
+
+    @api.depends('stock_valuation_layer_ids', 'product_id.lot_valuated')
+    @api.depends_context('to_date', 'company')
+    def _compute_value_svl(self):
```
Comment (amoyaux) [inline addons/stock_account/models/stock_lot.py:29]: image
It didn't trigger on update standard_price

Comment (amoyaux) [reply addons/stock_account/models/stock_lot.py:29]: image
It didn't trigger at all

**Comment by amoyaux** [comment]
image

File: addons/stock_account/models/stock_valuation_layer.py [hunk 11]
```diff
@@ -34,6 +35,7 @@ class StockValuationLayer(models.Model):
     reference = fields.Char(related='stock_move_id.reference')
     price_diff_value = fields.Float('Invoice value correction with invoice currency')
     warehouse_id = fields.Many2one('stock.warehouse', string="Receipt WH", compute='_compute_warehouse_id', search='_search_warehouse_id')
+    lot_id = fields.Many2one('stock.lot', 'Lot/Serial Number', check_company=True)
```
Comment (amoyaux) [inline addons/stock_account/models/stock_valuation_layer.py:38]: Don't forget to index it

File: addons/stock_account/models/product.py [hunk 12]
```diff
@@ -72,6 +110,100 @@ def write(self, vals):
     # -------------------------------------------------------------------------
     # Misc.
     # -------------------------------------------------------------------------
+    def _enable_lot_valuated(self):
+        self.ensure_one()
```
[the remark this answers is not in the document]
Comment (Whenrow) [reply addons/stock_account/models/product.py:109]: todo same as change product category

File: addons/stock_account/models/product.py [hunk 13]
```diff
@@ -262,85 +397,26 @@ def _change_standard_price(self, new_price):
             }
             svl_vals_list.append(svl_vals)
         stock_valuation_layers = self.env['stock.valuation.layer'].sudo().create(svl_vals_list)
+        stock_valuation_layers._change_standart_price_accounting_entries(new_price)
 
-        # Handle account moves.
-        product_accounts = {product.id: product.product_tmpl_id.get_product_accounts() for product in self}
```
Comment (amoyaux) [inline addons/stock_account/models/product.py:267]: check if it's not a dup of _validate_accounting_entries

File: addons/stock_account/models/product.py [hunk 14]
```diff
@@ -262,85 +397,26 @@ def _change_standard_price(self, new_price):
             }
             svl_vals_list.append(svl_vals)
         stock_valuation_layers = self.env['stock.valuation.layer'].sudo().create(svl_vals_list)
+        stock_valuation_layers._change_standart_price_accounting_entries(new_price)
```
Comment (amoyaux) [inline addons/stock_account/models/product.py:400]: check if it's not a dup of _validate_accounting_entries

File: addons/stock_account/models/stock_lot.py [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+            lot.avg_cost = avg_cost
+            lot.total_value = avg_cost * quantity_sum
+
+    def write(self, vals):
+        if 'standard_price' in vals and not self.env.context.get('disable_auto_svl'):
+            self._change_standard_price(vals['standard_price'])
+        return super().write(vals)
+
+    def _change_standard_price(self, new_price):
+        """Helper to create the stock valuation layers and the account moves
```
Comment (amoyaux) [inline addons/stock_account/models/stock_lot.py:70]: I think we sould not support it in standard

File: addons/stock_account/static/src/fields/boolean_confirm.js [hunk 16]
```diff
@@ -0,0 +1,59 @@
+/** @odoo-module **/
```
Comment (Whenrow) [inline addons/stock_account/static/src/fields/boolean_confirm.js:1]: @cth-odoo Do you think It could a standard feature ?
tl;dr: the same idea as `confirm=message` on a `<button>` but for a field. To be sure the value is actually changed only when the user click the confirm button of the modal. Thanks

File: addons/stock_account/models/product.py [hunk 17]
```diff
@@ -111,8 +215,33 @@ class ProductProduct(models.Model):
 
     def write(self, vals):
         if 'standard_price' in vals and not self.env.context.get('disable_auto_svl'):
-            self.filtered(lambda p: p.cost_method != 'fifo')._change_standard_price(vals['standard_price'])
-        return super(ProductProduct, self).write(vals)
+            self.filtered(lambda p: p.cost_method != 'fifo' and not p.lot_valuated)._change_standard_price(vals['standard_price'])
```
Comment (amoyaux) [inline addons/stock_account/models/product.py:218]: move the logic directly in _change_standard_price

File: addons/stock_account/models/product.py [hunk 18]
```diff
@@ -162,7 +291,7 @@ def action_revaluation(self):
     # -------------------------------------------------------------------------
     # SVL creation helpers
     # -------------------------------------------------------------------------
-    def _prepare_in_svl_vals(self, quantity, unit_cost):
+    def _prepare_in_svl_vals(self, quantity, unit_cost, lot_id=False):
```
Comment (amoyaux) [inline addons/stock_account/models/product.py:294]: lot if recordset.
lot_id if id
change the method or the arg passed to it (seems we don't need the record)

File: addons/stock_account/views/stock_valuation_layer_views.xml [hunk 19]
```diff
@@ -146,6 +151,7 @@
                 <filter string="Has Remaining Qty" name="has_remaining_qty" domain="[('remaining_qty', '>', 0)]"/>
                 <group expand='0' string='Group by...'>
                     <filter string='Product' name="group_by_product_id" context="{'group_by': 'product_id'}"/>
+                    <filter string='Lot/Serial number' name="group_by_lot_id" context="{'group_by': 'lot_id'}"/>
```
Comment (Felicious) [inline addons/stock_account/views/stock_valuation_layer_views.xml:154]: Hi @Whenrow (: Amazing job on this massive new feature! 

I'm a technical writer from the SF office. I'm not sure if it's my place to point this out, but I was wondering if we could capitalize "Number" here, to maintain consistency in capitalization with existing filters in this drop-down menu.

image
```suggestion
                    <filter string='Lot/Serial Number' name="group_by_lot_id" context="{'group_by': 'lot_id'}"/>
```

File: addons/stock_account/models/product.py [hunk 20]
```diff
@@ -113,7 +142,29 @@ class ProductProduct(models.Model):
...
+            return {
+                'warning': {
+                    'title': _("Warning"),
+                    'message': _("This product is valuated by lot/serial number. Changing the cost \
+will update the cost of every lot/serial number in stock."),
+                }
+            }
+
+    @api.onchange('lot_valuated')
+    def _onchange_lot_valuated(self):
```
Comment (amoyaux) [inline addons/stock_account/models/product.py:159]: Strange to have it in product.product but not in template while the field is in template

File: addons/stock_account/models/product.py [hunk 21]
```diff
@@ -646,27 +660,71 @@ def _svl_empty_stock(self, description, product_category=None, product_template=
...
+            ('product_id', 'in', self.product_variant_ids.ids),
+            ('lot_id', '!=', False),
+            ], ['product_id', 'location_id', 'lot_id'], ['quantity:sum'],
+            having=[('quantity:sum', '<', 0)])
+        lots = self.env['stock.quant']._read_group([
+            ('product_id', 'in', self.product_variant_ids.ids),
+            ('lot_id', '!=', False),
+            ], ['product_id', 'location_id', 'lot_id'], ['quantity:sum'],
+            having=[('quantity:sum', '>', 0)])
+        for group in lots:
```
Comment (amoyaux) [inline addons/stock_account/models/product.py:699]:
```suggestion
        for product, location, lot, quantity in lots:
```

File: addons/stock_account/wizard/stock_valuation_layer_revaluation.py [hunk 22]
```diff
@@ -15,6 +17,10 @@ class StockValuationLayerRevaluation(models.TransientModel):
     def default_get(self, default_fields):
         res = super().default_get(default_fields)
         context = self.env.context
+        if res.get('lot_id'):
+            lot = self.env['stock.lot'].browse(res['lot_id'])
+            if lot:
```
Comment (amoyaux) [inline addons/stock_account/wizard/stock_valuation_layer_revaluation.py:22]: .exists()

**Comment by amoyaux** [comment]
robodoo r+

**Comment by Whenrow** [comment]
robodoo r+
